### ISSUE TITLE
worldWind error in android level 23

### ISSUE THREAD

**Issue by shazde2025** [NASAWorldWind/WorldWindAndroid#62]
hi guys 
i use worldwind sdk and when i run my app earth shape is incorrect
but i run app on android level 19 work good
i think open Gl library dont work correct in android level 23
tanks for help
![photo_2016-09-04_21-28-04](https://cloud.githubusercontent.com/assets/11379106/18232539/945edc1c-72e6-11e6-86c0-9a846768df44.jpg)

**Comment by pdavidc** [comment]
Hello, and thanks for reaching out to us. We're not seeing this problem with Android SDK 23, and need more information about your environment. Can you send us the device make and model, Android OS version, and log output?

**Comment by shazde2025** [comment]
hi
my device is samsung galaxy note 4 and then i update to android 6 open gl
library dont work correct
my device has cpu  exynos 5433 - mail t760
i think this problem for update 6 for may device

**Comment by SimoneMike** [comment]
I am also seeing the same issue on the Panasonic FZ-B2 tablets running SDK 23. The version is 6.0.1 running on the tablet.

**Comment by jgiovino** [comment]
I have seen similar issues with past experience where we exceed the bounds of the shader's float data type. The image above shows a lot of ones and zeros in the vertex coordinates. Please verify when updating OS's the GLES stack did not get downgraded to a SW renderer. It would be interesting to see both "OpenGL driver version" under "About tablet" before and after upgrade. One way to force this type of behavior is to run the application on the AVD and make sure to NOT select "Use Host GPU". BIG IF, this is in fact the reason THEN the solution is to normalize the world coordinate system prior to passing vertices down the GL pipeline. In other words, change the world units from meters to kilometers or some multiple thereof.

**Comment by Skalaw** [comment]
Also problem with LG-K350 API 23

**Comment by Sufaev** [comment]
I have Samsung Galaxy Note 5 and Samsung Galaxy Tab S 8.4 with the same bug.
I think most of Android 6.0 devices should have this bug.

**Comment by Sufaev** [comment]
It looks like this problem is related to all Samsung devices on Android 23.

**Comment by zglueck** [comment]
@jgiovino would utilizing the ```highp``` float precision type in the vertex shader help with the situation?

**Comment by Andro-Samurai** [comment]
![1bug](https://cloud.githubusercontent.com/assets/25287348/22186916/46b39852-e106-11e6-9632-947a6d68968d.png)

the screenshot of GALAXY S6 - ANDROID 5.1 

Also there it appears in:
1- DEVICE
Device Galaxy J7(2016) (j7xelte) Manufacturer Samsung Device type Phone Device language Polish CPU make Samsung CPU model Exynos 7870 Native platform armeabi-v7a, armeabi RAM (MB) 2048 Screen size 720 × 1280 Screen density (dpi) 320 OpenGL ES version 3.0 OS Android 6.0

2- DEVICE
Device Galaxy S7 (herolte) Manufacturer Samsung Device type Phone Device language English CPU make Samsung CPU model Exynos 8890 Native platform armeabi-v7a, armeabi, arm64-v8a RAM (MB) 4096 Screen size 1440 × 2560 Screen density (dpi) 640 OpenGL ES version 3.1 OS Android 6.0

3- DEVICE
Device A3-A40 (acer_jetfirefhd) Manufacturer Acer Device type Tablet Device language English CPU make Mediatek CPU model MT8163 Native platform armeabi-v7a, armeabi, arm64-v8a RAM (MB) 2048 Screen size 1920 × 1200 Screen density (dpi) 240 OpenGL ES version 3.0 OS Android 6.0

4- DEVICE
Device Galaxy TabS 10.5 (chagalllte) Manufacturer Samsung Device type Tablet Device language English CPU make Samsung CPU model Exynos 5420 Native platform armeabi-v7a, armeabi RAM (MB) 3072 Screen size 2560 × 1600 Screen density (dpi) 320 OpenGL ES version 3.0 OS Android 6.0

5- DEVICE
Device Galaxy S6 (zeroflteatt) Manufacturer Samsung Device type Phone Device language English CPU make Samsung CPU model Exynos 7420 Native platform armeabi-v7a, armeabi, arm64-v8a RAM (MB) 3072 Screen size 1440 × 2560 Screen density (dpi) 640 OpenGL ES version 3.1 OS Android 6.0

6- DEVICE
Device Galaxy S7 Edge (hero2lte) Manufacturer Samsung Device type Phone Device language English CPU make Samsung CPU model Exynos 8890 Native platform armeabi-v7a, armeabi, arm64-v8a RAM (MB) 4096 Screen size 1440 × 2560 Screen density (dpi) 640 OpenGL ES version 3.1 OS Android 6.0

7- DEVICE
Device Lenovo A2016b30 (A2016b30) Manufacturer Lenovo Mobile Device type Phone Device language Portuguese CPU make Mediatek CPU model MT6735M Native platform armeabi-v7a, armeabi RAM (MB) 1024 Screen size 480 × 854 Screen density (dpi) 240 OpenGL ES version 3.0 OS Android 6.0

8- DEVICE
Device Galaxy S5 Neo (s5neoltecan) Manufacturer Samsung Device type Phone Device language English CPU make Samsung CPU model Exynos 7580 Native platform armeabi-v7a, armeabi RAM (MB) 2048 Screen size 1080 × 1920 Screen density (dpi) 480 OpenGL ES version 3.0 OS Android 6.0

**Comment by jgiovino** [comment]
These images do in fact look like the GPU float has been exceeded.

You can try highp in both vert and frag shaders but since its happening on this many devices it probably is not the issue.

I have been able to normalize world coordinates on SW renderers using the previous 'WorldDroid' WorldWind implementation from Trilogis.

We need to track down the GPU chipset for the devices below and see if its failing on 'like technology' then we can track down that graphics stack.

Bottom line, for each of the 8 devices listed lets figure out the GPU technology backing it.

If they are in fact on different SoC's then we need to see if they are sharing common GPU core logic and/or common dri/drm drivers.

v/r,

Jeff

________________________________

**Comment by Andro-Samurai** [comment]
@jgiovino I'v tracked down the GPU chipset of the devices
All of them have a Mali series GPU, to be exact Mali-T(number)
Please check them  out https://en.wikipedia.org/wiki/Mali_(GPU)#Variants 
do you notice anything?

**Comment by jgiovino** [comment]
Interesting read…
https://community.arm.com/graphics/b/blog/posts/benchmarking-floating-point-precision-in-mobile-gpus
it seems to focus on precision related to the fragment shader and not the vertex which is where are problem seems to lie,.
http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.dui0363d/CJAFCCDE.html
this does not explicitly state the floating point architecture for the vertex shader only implies by stating “the geometry processor”

Looking at the problem backwards we can calculate the number of bits for the exponent is required to represent world coordinates defined as meters.

Value = (-1)^s * (1+sig) * 2^(exb-bias) assume zero bias

6376136 = 2 * 2^e = 2^(e+1)
e > 21
that only requires 7 exponent bits in twos compliment

So maybe the question is: Do the Mali GPUs have 7 or less exponent bits in their vertex shader’s lowp floating point numbers?
This doesn’t seem to pass the sniff test does it?

**Comment by Sufaev** [comment]
Devices with the same GPU works normally on previous Android. So I think it is not a hardware issue of chip.

**Comment by jgiovino** [comment]
So somewhere between the pipeline and the GPU our vertecies are being mushed.
That seems to now point to a driver, or reverting to a SW stack.
I wonder if specifying gles version in the shaders could help.
#version 100

**Comment by Andro-Samurai** [comment]
Even though most of affected devices are running API 23 but there is an incident that it did not work properly on API 22.

Also I found out that it worked on the same device if the CPU was different; for example: 
it did not work correctly with **Galaxy S7 Edge (hero2lte) CPU make: Samsung, CPU model: Exynos 8890, API: 23**  but worked perfect with  **Galaxy S7 Edge (hero2qltevzw) CPU make: Qualcomm, CPU model: MSM8996, API: 24**

**Comment by Andro-Samurai** [comment]
Hello Guys,

I think this bug is the most important one to fix because it affects nearly 50% of the people because the affected devices are (unfortunately) the most popular devices by far.
could you please tag this as "important" or "Urgent"

you can perform tests on real devices from Samsung remote test lab 
[http://developer.samsung.com/remotetestlab](url)

Thank you for your help

**Comment by pdavidc** [comment]
@Andro-Samurai You're absolutely right, and we're working on a fix. See #78, which is more general and we think captures the actual underlying cause.

**Comment by pdavidc** [comment]
@Andro-Samurai, @shazde2025, @Sufaev, @SimoneMike, @Skalaw : Would you please try the latest code in the develop branch, and let us know whether or not the issue is fixed?

**Comment by Sufaev** [comment]
Partially fixed. I think Atmosphere shader not fixed.

![screenshot_2017-01-30-21-17-50](https://cloud.githubusercontent.com/assets/12487306/22437707/c6e22c6a-e731-11e6-9bcd-82cab50e963c.png)

**Comment by Sufaev** [comment]
Here is wothout atmosphere layer. Samsung Tab S 8.4 Android 23
![screenshot_2017-01-30-21-21-54](https://cloud.githubusercontent.com/assets/12487306/22437835/413b915e-e732-11e6-90bc-12f2afe419b4.png)

**Comment by Sufaev** [comment]
It works now  Samsung Tab S 8.4 Android 23.
![screenshot_2017-01-30-21-34-09](https://cloud.githubusercontent.com/assets/12487306/22438241/e8656436-e733-11e6-8935-22413309a7fd.png)

**Comment by SimoneMike** [comment]
David,

The fix worked beautifully. Thank you for taking care of this issue. 

-Mike
